File: BowlingGame/ObjC/BowlingGame/Frame.m
//
//  Frame.m
//  BowlingGame
//
//  Created by Tony Hoang on 4/16/18.
//  Copyright © 2018 me. All rights reserved.
//

#import "Frame.h"
#import "Roll.h"
#import "Game.h"

@interface Frame()

@property (nonatomic, strong) NSMutableArray* rewards;
@property (nonatomic, assign, readonly) NSInteger baseScore;

@property (nonatomic, assign, readonly) BOOL isStrike;
@property (nonatomic, assign, readonly) BOOL isSpare;

@end

const NSInteger MAX_SCORE_PER_FRAME = 10;
const NSInteger MAX_ROLLS_PER_FRAME = 2;

@implementation Frame

- (instancetype)initWithRoll:(Roll*)roll {
    if (self = [super init]) {
        _rolls = [NSMutableArray new];
        _rewards = [NSMutableArray new];
        
        [_rolls addObject:roll];
    }
    
    return self;
}

- (NSInteger)score {
    NSInteger score = 0;
    
    score += self.baseScore;
    
    for (NSNumber* reward in self.rewards) {
        score += reward.integerValue;
    }
    
    return score;
}

- (NSInteger)baseScore {
    NSInteger score = 0;
    
    for (Roll* roll in self.rolls) {
        score += roll.score;
    }

    return score;
}

- (BOOL)isStrike {
    Roll* firstRoll = self.rolls.firstObject;
    return firstRoll.score == MAX_SCORE_PER_FRAME;
}

- (BOOL)isSpare {
    return ![self isStrike] && (self.baseScore == MAX_SCORE_PER_FRAME);
}

- (BOOL)isFinished {
    return self.isStrike || self.rolls.count == MAX_ROLLS_PER_FRAME;
}

- (void)addBonusFromRoll:(Roll*)roll {
    if (self.isStrike) {
        if (self.rewards.count < NUM_OF_BONUS_ROLLS_PER_STRIKE) {
            [self.rewards addObject:@(roll.score)];
        }
    } else if (self.isSpare) {
        if (self.rewards.count < NUM_OF_BONUS_ROLLS_PER_SPARE) {
            [self.rewards addObject:@(roll.score)];
        }
    }
}

- (NSString*)description {
    NSString* desc = @"(rolls: (";
    
    for (Roll* roll in self.rolls) {
        desc = [desc stringByAppendingFormat:@"%ld, ", roll.score];
    }
    
    desc = [desc stringByAppendingString:@"), bonuses: ("];
    
    for (NSNumber* reward in self.rewards) {
        desc = [desc stringByAppendingFormat:@"%ld, ", reward.integerValue];
    }
    
    desc = [desc stringByAppendingString:@"))"];
    
    return desc;
}

@end

File: BowlingGame/Swift/BowlingGame/Frame.swift
//
//  Frame.swift
//  BowlingGame
//
//  Created by Tony Hoang on 4/16/18.
//  Copyright © 2018 me. All rights reserved.
//

import Foundation

class Frame: CustomStringConvertible, Equatable {
    var rolls = [Roll]()
    var bonuses = [Int]()
    
    init(roll: Roll) {
        rolls.append(roll)
    }

    static func == (lhs: Frame, rhs: Frame) -> Bool {
        return lhs.rolls == rhs.rolls
    }
    
    func updateBonus(from roll:Roll) {
        if (self.isStrike) {
            if (bonuses.count < Game.numOfBonusRollPerStrike) {
                bonuses.append(roll.score)
            }
        } else if (self.isSpare) {
            if (bonuses.count < Game.numOfBonusRollPerSpare) {
                bonuses.append(roll.score)
            }
        }
    }
    
    var description: String {
        var desc = "rolls: ("
        desc += rolls.compactMap({String($0.score)}).joined(separator: ", ")
    
        desc += "), bonuses: ("
        desc += bonuses.compactMap({ String($0)}).joined(separator: ", ")
        desc += "))"

        return desc
    }
    
    var isStrike: Bool {
        return rolls[0].score == Game.maxScorePerFrame
    }
    
    var isSpare: Bool {
        return !isStrike && (score == Game.maxScorePerFrame)
    }
    
    var isFinished: Bool {
        return isStrike || isSpare || rolls.count == Game.maxRollPerFrame
    }
    
    var score: Int {
        return rolls.compactMap({Int($0.score)}).reduce(0, +) + bonuses.reduce(0, +)
    }
}
